File: share/share-client/share-client-hbase/src/main/scala/com/hbase/HbaseUtil.scala
package com.hbase

import java.text.SimpleDateFormat
import java.util.Date

import com.alibaba.fastjson.JSONObject
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.hbase.client._
import org.apache.hadoop.hbase.filter.{CompareFilter, RowFilter, SubstringComparator}
import org.apache.hadoop.hbase.io.ImmutableBytesWritable
import org.apache.hadoop.hbase.mapred.{TableInputFormat, TableOutputFormat}
import org.apache.hadoop.hbase.util.Bytes
import org.apache.hadoop.hbase.{HBaseConfiguration, TableName}
import org.apache.hadoop.mapred.JobConf
import org.apache.log4j.Logger
import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD

import scala.collection.JavaConversions._
import scala.collection.mutable.ArrayBuffer

object HbaseUtil {
  @transient lazy val logger: Logger = Logger.getLogger(this.getClass().getName())
  //参数
  val tableName: String = "orderTable"
  val quorum = "ddx,ddy,ddz"//gaiaa,gaiab,gaiac,gaiad,gaiae
  val master = "ddx"
  /**
    * @function 获取hbase配置
    * @return
    */
  def getHbaseConf(): Configuration = {
    var hbaseConf: Configuration = null
    try {
      //获取hbase的conf
      hbaseConf = HBaseConfiguration.create()
      //设置写入的表
      hbaseConf.set("hadoop.home.dir", "/d1/bin/hadoop")
      hbaseConf.set("hbase.zookeeper.property.clientPort", "2181")
      hbaseConf.set("hbase.zookeeper.quorum", quorum) // 47.97.206.57,47.97.183.128,47.96.3.154,47.99.131.49,47.98.106.110
      hbaseConf.set("hbase.master", s"${master}:60010")
      hbaseConf.set("hbase.client.pause", "500")
      hbaseConf.set("hbase.client.retries.number", "500")
      hbaseConf.set("hbase.rpc.timeout", "20000000")
      hbaseConf.set("hbase.client.operation.timeout", "300000")
      hbaseConf.set("hbase.client.scanner.timeout.period", "100000")
    } catch {
      case e: Exception => logger.error(">>>连接hbase失败:," + e)
    }
    hbaseConf
  }

  /**
    * @function 获得操作类Table
    * @param tableName
    * @return
    */
  def getTable(tableName: String): Table = {
    var table: Table = null
    try {
      val hbaseConf = getHbaseConf()
      val conn = ConnectionFactory.createConnection(hbaseConf)
      table = conn.getTable(TableName.valueOf(tableName))
    } catch {
      case e: Exception => logger.error(">>>获取Table对象失败:" + e)
    }
    table
  }

  /**
    * @function 添加一行数据
    * @param table  操作表的对象
    * @param rowKey key
    * @param family 簇
    * @param column 列
    * @param value  需要添加的数据内容
    */
  def addRow(table: Table, rowKey: String, family: String, column: String, value: String): Unit = {
    val rowPut: Put = new Put(Bytes.toBytes(rowKey))
    if (value == null) {
      rowPut.addColumn(family.getBytes, column.getBytes, "".getBytes)
    } else {
      rowPut.addColumn(family.getBytes, column.getBytes, value.getBytes)
    }
    table.put(rowPut)
  }


  /**
    * @function rdd数据入到hbase库
    * @param inputRdd 需要入库的RDD
    */
  def putRdd(inputRdd: RDD[JSONObject]): Unit = {
    try {
      val jobConf = new JobConf(getHbaseConf())
      jobConf.setOutputFormat(classOf[TableOutputFormat])
      jobConf.set(TableOutputFormat.OUTPUT_TABLE, tableName)
      val putRdd = inputRdd.map(convert)
      putRdd.saveAsHadoopDataset(jobConf)
    } catch {
      case e: Exception => logger.error(">>>数据入库失败:" + e)
    }
  }

  /**
    * @function 构建转换数据
    * @param obj
    * @return
    */
  def convert(obj: JSONObject) = {
    var put: Put = null
    try {
      //设置主键，这里测试，使用系统时间
      val longtTime = System.currentTimeMillis()
      val dfTime = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss:SSS")
      val saveTime = dfTime.format(new Date(longtTime))
      val key = saveTime
      put = new Put(Bytes.toBytes(key))
      put.addColumn(Bytes.toBytes("info"), Bytes.toBytes("date"), Bytes.toBytes(saveTime))
      put.addColumn(Bytes.toBytes("info"), Bytes.toBytes("wrong_value"), Bytes.toBytes(obj.toString))
    } catch {
      case e: Exception => println(">>>数据转换失败：" + e)
    }
    (new ImmutableBytesWritable, put)
  }

  /**
    * @function 根据rowKey精确查询单行数据单行值
    * @param rowkey
    * @param family
    * @param column
    * @return
    */
  def getByKey(table: Table, rowkey: String, family: String, column: String): String = {
    var value: String = null
    try {
      //单个get查询
      val get = new Get(Bytes.toBytes(rowkey))
      get.addColumn(Bytes.toBytes(family), Bytes.toBytes(column))
      val result = table.get(get)
      value = Bytes.toString(result.getValue(Bytes.toBytes(family), Bytes.toBytes(column)))
    } catch {
      case e: Exception => logger.error(">>>查询失败:" + e)
    }
    value
  }

  /**
    * @function 根据rowKey精确查询单行数据的多列值
    * @param rowkey
    * @param family
    * @param columns
    * @return
    */
  def getByKey1(table: Table, rowkey: String, family: String, columns: Array[String]): ArrayBuffer[String] = {
    var list: ArrayBuffer[String] = new ArrayBuffer[String]()
    try {
      val get = new Get(Bytes.toBytes(rowkey))
      get.addFamily(family.getBytes())
      val result = table.get(get)
      columns.foreach(column => {
        val value = Bytes.toString(result.getValue(Bytes.toBytes(family), Bytes.toBytes(column)))
        list += value
      })
    } catch {
      case e: Exception => logger.error(">>>查询失败:" + e)
    }
    list
  }

  /**
    * @function 根据subKey模糊批量查询多行数据
    * @param subKey
    * @param family
    * @param column
    * @return
    */
  def scanBySubKey(table: Table, subKey: String, family: String, column: String): ArrayBuffer[String] = {
    val resultList = new ArrayBuffer[String]()
    val table = getTable(tableName)
    //批量模糊查询
    val filter = new RowFilter(CompareFilter.CompareOp.EQUAL, new SubstringComparator(subKey))
    val scan = new Scan()
    scan.setFilter(filter)
    val rs: ResultScanner = table.getScanner(scan)
    for (result <- rs) {
      if (result.containsColumn(Bytes.toBytes(family), Bytes.toBytes(column))) {
        val value = Bytes.toString(result.getValue(Bytes.toBytes(family), Bytes.toBytes(column)))
        resultList += value
      }
    }
    resultList
  }

  /**
    * @function 根据rowKey删除整行
    * @param table
    * @param rowKey
    */
  def deleteByKey(table: Table, rowKey: String): Unit = {
    try {
      table.delete(new Delete(rowKey.getBytes()))
    } catch {
      case e: Exception => println(">>>删除操作失败：" + e)
    }
  }

  /**
    * @function 删除rowKey的某簇的某列
    * @param table
    * @param rowKey
    */
  def deleteByColumn(table: Table, rowKey: String, family: String, column: String): Unit = {
    try {
      val delete = new Delete(rowKey.getBytes())
      delete.addColumn(family.getBytes(), column.getBytes())
      table.delete(delete)
    } catch {
      case e: Exception => println(">>>删除操作失败：" + e)
    }
  }
  /**
    * @function 删除rowKey的某个family
    * @param table
    * @param rowKey
    */
  def deleteByFamily(table: Table, rowKey: String, family: String): Unit = {
    try {
      val delete = new Delete(rowKey.getBytes())
      delete.addFamily(family.getBytes())
      table.delete(delete)
    } catch {
      case e: Exception => println(">>>删除操作失败：" + e)
    }
  }

  /**
    * @function 根据rowKey模糊删除
    * @param table
    * @param subKey
    */
  def deleteByKeys(table: Table, subKey: String): Unit = {
    try {
      //批量模糊删除
      val filter = new RowFilter(CompareFilter.CompareOp.EQUAL, new SubstringComparator(subKey))
      val scan = new Scan()
      scan.setFilter(filter)
      val scanners: ResultScanner = table.getScanner(scan)
      val keyList = new ArrayBuffer[Delete]()
      var delete: Delete = null
      //注意：scala和java的for循环有区别，需要引入转换  import scala.collection.JavaConversions._
      for (scanner <- scanners) {
        val rowKey = new String(scanner.getRow)
        delete = new Delete(rowKey.getBytes)
        keyList += delete
      }
      table.delete(keyList)
    } catch {
      case e: Exception => println(">>>删除操作失败：" + e)
    }
  }
  /**
    * @function 读取hbase数据
    * @param servers
    * @param sc
    * @param table
    */
//  def loadHbase(servers: String, sc: SparkContext, table: String) = {
//    val hbaseConf = HBaseConfiguration.create()
//    hbaseConf.set("hbase.zookeeper.quorum", servers)
//    hbaseConf.set("hbase.zookeeper.property.clientPort", "2181")
//    hbaseConf.set("zookeeper.session.timeout", "6000000")
//    hbaseConf.set(TableInputFormat.INPUT_TABLE, table)
//    val ratingsData = sc.newAPIHadoopRDD(hbaseConf, classOf[TableInputFormat], classOf[ImmutableBytesWritable], classOf[Result])
//    val hbaseRatings = ratingsData.map { case (_, res) =>
//      val app_user_id = Bytes.toString(res.getValue(Bytes.toBytes("info"), Bytes.toBytes("app_user_id")))
//      val extra_id = Bytes.toString(res.getValue(Bytes.toBytes("info"), Bytes.toBytes("extra_id")))
//      val count = Bytes.toString(res.getValue(Bytes.toBytes("info"), Bytes.toBytes("count")))
//      Array(extra_id.toInt, count.toDouble).mkString("\t")
//    }
//  }

}

File: share/share-client/share-client-hbase/src/main/java/com/hbase/HBaseUtil.java
package com.hbase;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.hbase.*;
import org.apache.hadoop.hbase.client.*;
import org.apache.hadoop.hbase.util.Bytes;
import org.apache.log4j.Logger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class HBaseUtil {
    private static Configuration conf;
    private static Logger logger = Logger.getLogger(HBaseUtil.class);

    static {
        conf = HBaseConfiguration.create();
        conf.set("hbase.rootdir", "hdfs://datanode:9000/hbase");
        conf.set("hbase.zookeeper.quorum", "datanode:2181");
    }

    /**
     * @param tableName
     * @return
     * @throws IOException
     * @function 判断表是否存在
     */
    public static boolean isExist(String tableName) throws IOException {
        logger.info("正在调用isExist()方法，判断" + tableName + "是否存在");
        //新API
        Connection connection = ConnectionFactory.createConnection(conf);       //工厂设计模式
        Admin admin = connection.getAdmin();
        return admin.tableExists(TableName.valueOf(tableName));
    }

    /**
     * @param tableName
     * @param columnFamily
     * @throws Exception
     * @function Hbase表的创建
     */
    public static void createTable(String tableName, String... columnFamily) throws Exception {
        logger.info("正在调用createTable()方法创建表" + tableName);
        Connection connection = ConnectionFactory.createConnection(conf);       //工厂设计模式
        Admin admin = connection.getAdmin();
        //判断表是否存在
        if (isExist(tableName)) {
            throw new Exception("表已存在，不可重复创建");
        } else {
            //创建表描述器
            HTableDescriptor desc = new HTableDescriptor(TableName.valueOf(tableName));
            for (String family : columnFamily) {
                desc.addFamily(new HColumnDescriptor(family));
            }
            //创建表
            admin.createTable(desc);
            logger.info(tableName + "表成功创建");
        }
    }

    /**
     * @param map 此时用Map集合来封装表名称和列族 HashMap<String,String[]> key为tableName,value为String[]数组的columnFamily
     * @throws Exception
     * @function 批量创建表
     */
    public static void createMultiTables(Map<String, String[]> map) throws Exception {
        logger.info("正在调用createMultiTable()创建表");
        for (Map.Entry<String, String[]> entry : map.entrySet()) {
            //1. 获取tableName和columnFamily
            //1.1 获取tableName
            String tableName = entry.getKey();
            //1.2 获取columnFamily
            String[] columnFamily = entry.getValue();
            //创建表
            createTable(tableName, columnFamily);
        }
    }


    /**
     * @param tableName
     * @throws Exception
     * @function Hbase表的删除
     */
    public static void deleteTable(String tableName) throws Exception {
        logger.info("正在调用deleteTable删除表" + tableName);
        Connection connection = ConnectionFactory.createConnection(conf);       //工厂设计模式
        Admin admin = connection.getAdmin();
        if (isExist(tableName)) {
            if (!admin.isTableDisabled(TableName.valueOf(tableName))) {
                //disable
                admin.disableTable(TableName.valueOf(tableName));
            }
            //删除表
            admin.deleteTable(TableName.valueOf(tableName));
            logger.info(tableName + "表删除成功=======");
        } else {
            throw new Exception(tableName + "表不存在，无法进行删除");
        }
    }

    /**
     * @param list List<String>中封装tableName集合
     * @throws Exception
     * @function 批量删除表
     */
    public static void deleteMultiTables(List<String> list) throws Exception {
        for (String tableName : list) {
            deleteTable(tableName);
        }
    }

    /**
     * @param tableName    表名
     * @param rowKey       行键
     * @param columnFamily 列族
     * @param column       列名
     * @param value        列值
     * @throws IOException
     * @function 向Hbase中添加一行数据
     */
    public static void addRow(String tableName, String rowKey, String columnFamily, String column, String value) throws IOException {
        logger.info("正在调用addRow方法，向" + tableName + "表中添加数据");
        //工厂设计模式
        Connection connection = ConnectionFactory.createConnection(conf);
        //获取表
        Table table = connection.getTable(TableName.valueOf(tableName));
        //封装待存放的对象rowkey封装
        Put put = new Put(Bytes.toBytes(rowKey));
        put.addColumn(Bytes.toBytes(columnFamily), Bytes.toBytes(column), Bytes.toBytes(value));      //指定列族，列名，存放的数据
        //添加数据
        table.put(put);     //可以传入Put对象的集合
    }

    /**
     * @function 批量添加数据
     * @param tableName
     * @param puts
     * @throws Exception
     */
    public static void addRows(String tableName, List<Put> puts) throws Exception {
        Connection connection = ConnectionFactory.createConnection(conf);
        HTable htable = (HTable) connection.getTable(TableName.valueOf(tableName));
        htable.setAutoFlushTo(false);
        htable.setWriteBufferSize(5 * 1024 * 1024);
        try {
            htable.put((List<Put>) puts);
            htable.flushCommits();
        } finally {
            if (htable != null)
                htable.close();
            connection.close();
        }
    }


    /**
     * @param tableName
     * @param rowKey
     * @throws IOException
     * @function 删除单行数据
     */
    public static void deleteMultiRow(String tableName, String rowKey) throws IOException {
        logger.info("正在调用deleteMultiRow方法，删除" + tableName + "表中rowKey为" + rowKey + "的数据");
        Connection connection = ConnectionFactory.createConnection(conf);       //工厂设计模式
        Table table = connection.getTable(TableName.valueOf(tableName));        //获取表
        Delete delete = new Delete(Bytes.toBytes(rowKey));
        table.delete(delete);
    }

    /**
     * @param tableName
     * @param rowKeys
     * @throws IOException
     * @function 删除多行数据
     */
    public static void deleteMultiRow(String tableName, List<String> rowKeys) throws IOException {
        logger.info("正在调用deleteMultiRow方法，删除" + tableName + "表中" + "多行数据");
        Connection connection = ConnectionFactory.createConnection(conf);       //工厂设计模式
        //获取表
        Table table = connection.getTable(TableName.valueOf(tableName));
        //批量删除
        List<Delete> list = new ArrayList();
        for (String rowKey : rowKeys) {
            Delete delete = new Delete(Bytes.toBytes(rowKey));
            list.add(delete);
        }
        //批量删除多行
        table.delete(list);                                 //批处理效率更高
    }

    /**
     * @param tableName
     * @throws IOException
     * @function 扫描数据——得到所有数据
     */
    public static void getAllRows(String tableName) throws IOException {
        logger.info("正在调用getAllRows方法，扫描表" + tableName);
        Connection connection = ConnectionFactory.createConnection(conf);       //工厂设计模式
        //获取表
        Table table = connection.getTable(TableName.valueOf(tableName));
        Scan scan = new Scan();                     //扫描表的配置信息
//      scan.setMaxVersions();
        ResultScanner scanner = table.getScanner(scan);
        for (Result result : scanner) {
//          System.out.println(Bytes.toString(result.getRow()));        也可以获取到row key
            Cell[] rawCells = result.rawCells();            //获取所有的cell
            for (Cell cell : rawCells) {
                System.out.println("行健：" + Bytes.toString(CellUtil.cloneRow(cell)));
                System.out.println("行健：" + Bytes.toString(CellUtil.cloneFamily(cell)));
                System.out.println("行健：" + Bytes.toString(CellUtil.cloneQualifier(cell)));
                System.out.println("行健：" + Bytes.toString(CellUtil.cloneValue(cell)));
                System.out.println("------------------");
            }
        }
    }

    /**
     * @param tableName
     * @param rowKey
     * @param columnFalily
     * @param column
     * @throws IOException
     * @function 得到指定列数据
     */
    public static void getRow(String tableName, String rowKey, String columnFalily, String column) throws IOException {
        logger.info("正在调用getRow方法，获取表" + tableName + "中，rowKey为" + rowKey + "的数据");
        Connection connection = ConnectionFactory.createConnection(conf);       //工厂设计模式
        //获取表
        Table table = connection.getTable(TableName.valueOf(tableName));
        Get get = new Get(Bytes.toBytes(rowKey));
        get.addColumn(Bytes.toBytes(columnFalily), Bytes.toBytes(column));   //设置要获取的列族和列
        Result result = table.get(get);
        Cell[] rawCells = result.rawCells();            //获取所有的cell
        for (Cell cell : rawCells) {
            System.out.println("行健：" + Bytes.toString(CellUtil.cloneRow(cell)));
            System.out.println("行健：" + Bytes.toString(CellUtil.cloneFamily(cell)));
            System.out.println("行健：" + Bytes.toString(CellUtil.cloneQualifier(cell)));
            System.out.println("行健：" + Bytes.toString(CellUtil.cloneValue(cell)));
            System.out.println("------------------");
        }
    }
}
